### PR TITLE
Removes requirement to render entire device Screen 

### PR DESCRIPTION
**Pull request by rharter**
This adds support for `RenderMode.SHRINK`, allowing you to make view-only screenshots without saving an entire device frame.

### CODE REVIEW

File: paparazzi/src/main/java/app/cash/paparazzi/Paparazzi.kt [hunk 1]
```diff
@@ -109,8 +110,8 @@ class Paparazzi @JvmOverloads constructor(
   private val contentRoot = """
```
Comment (liutikas) [inline paparazzi/src/main/java/app/cash/paparazzi/Paparazzi.kt:110]: I'd leave a comment about what this is and that it is dynamic.

File: paparazzi/src/main/java/app/cash/paparazzi/Paparazzi.kt [hunk 2]
```diff
@@ -372,7 +377,7 @@ class Paparazzi @JvmOverloads constructor(
   private fun scaleImage(image: BufferedImage): BufferedImage {
     val maxDimension = Math.max(image.width, image.height)
     val scale = THUMBNAIL_SIZE / maxDimension.toDouble()
-    return ImageUtils.scale(image, scale, scale)
+    return if (scale < 1f) ImageUtils.scale(image, scale, scale) else image
```
Comment (liutikas) [inline paparazzi/src/main/java/app/cash/paparazzi/Paparazzi.kt:380]: I'd leave a comment on why we don't upscale the image.

**Review by liutikas — COMMENTED** [review]
LGTM

File: paparazzi/src/main/java/app/cash/paparazzi/Paparazzi.kt (hunk 1 above)
Comment (rharter) [reply paparazzi/src/main/java/app/cash/paparazzi/Paparazzi.kt:114]: Fixed in [0d9a21a](https://github.com/cashapp/paparazzi/pull/497/commits/0d9a21ae563fec566ec2f93783fb61c7b9699eaa)

File: paparazzi/src/main/java/app/cash/paparazzi/Paparazzi.kt (hunk 2 above)
Comment (rharter) [reply paparazzi/src/main/java/app/cash/paparazzi/Paparazzi.kt:387]: Fixed in [4b688ce](https://github.com/cashapp/paparazzi/pull/497/commits/4b688ce774b03c6bd538962618e867312b56dfbc)

**Comment by rharter** [comment]
Since there were significant structural changes I've updated this PR based on the latest master branch.

Thanks for your feedback, @liutikas, I've resolved your comments since the files have moved, but I did add the requested comments.

**Comment by rharter** [comment]
@jrodbx Anything preventing this from going in?

**Comment by jrodbx** [comment]
It's on the queue for review, there's some other stuff I'd like to get in first + internal feature work to deliver.  Hope to have a look soon.

**Comment by liutikas** [comment]
For what's worth androidx would love this feature to make 1.1 as well

File: paparazzi/paparazzi/src/main/java/app/cash/paparazzi/Paparazzi.kt [hunk 3]
```diff
@@ -104,11 +104,16 @@ class Paparazzi @JvmOverloads constructor(
   val context: Context
     get() = RenderAction.getCurrentContext()
 
+  /**
+   * The root layout that test views will be placed into. The FrameLayout is dynamically set to
+   * `wrap_content` if the `renderMode` is `RenderingMode.SizeAction.SHRINK` in the appropriate
+   * direction, otherwise it is set to `match_parent`.
+   */
```
Comment (jrodbx) [inline paparazzi/paparazzi/src/main/java/app/cash/paparazzi/Paparazzi.kt:111]: nit: remove, the code is readable as is.

File: paparazzi/paparazzi/src/test/java/app/cash/paparazzi/RenderingModeTest.kt [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+import app.cash.paparazzi.internal.ImageUtils
+import com.android.ide.common.rendering.api.SessionParams
+import org.junit.Test
+import org.junit.runner.Description
+import java.awt.image.BufferedImage
+import java.io.File
+import javax.imageio.ImageIO
+
+class RenderingModeTest {
+
```
Comment (jrodbx) [inline paparazzi/paparazzi/src/test/java/app/cash/paparazzi/RenderingModeTest.kt:21]:
```suggestion
```

**Review by jrodbx — APPROVED** [review]

File: paparazzi/paparazzi/src/main/java/app/cash/paparazzi/Paparazzi.kt (hunk 3 above)
Comment (jrodbx) [inline paparazzi/paparazzi/src/main/java/app/cash/paparazzi/Paparazzi.kt:111]:
```suggestion
```

**Comment by jrodbx** [comment]
I'll follow up with a PR to migrate the test as a plugin test, but this is great.  thanks!

**Comment by JavierStuart** [comment]
Should this work with Compose?

**Comment by rharter** [comment]
We use it with Compose views without issues. If you're having a specific failure perhaps open an issue?

**Pull request merged**